java-checks/src/main/java/org/sonar/java/checks/spring/SpringAntMatcherOrderCheck.java

```diff
@@ -0,0 +1,185 @@
+      }
+    }
+  }
+
+  @VisibleForTesting
+  static boolean matches(String pattern, String text) {
+    if (pattern.equals(text)) {
+      return true;
+    }
+    if (pattern.endsWith("**") && text.startsWith(pattern.substring(0, pattern.length() - 2))) {
+      return true;
+    }
+    boolean antPatternContainsRegExp = pattern.contains("{");
+    boolean textIsAlsoAnAntPattern = MATCHER_SPECIAL_CHAR.matcher(text).find();
+    if (pattern.isEmpty() || antPatternContainsRegExp || textIsAlsoAnAntPattern) {
+      return false;
+    }
+    return text.matches(antMatcherToRegEx(pattern));
+  }
+
+  @VisibleForTesting
+  static String antMatcherToRegEx(String pattern) {
+    // Note, regexp is not supported: {spring:[a-z]+} matches the regexp [a-z]+ as a path variable named "spring"
+
+    // escape regexp special characters
+    return escapeRegExpChars(pattern)
+      // ? matches one character
+      .replace("?", "[^/]")
+      // ** matches zero or more directories in a path ("$$" is a temporary place holder)
+      .replace("**", "$$")
+      // * matches zero or more characters
+      .replace("*", "[^/]*")
+      .replace("$$", ".*");
+  }
+
+  @VisibleForTesting
+  static String escapeRegExpChars(String pattern) {
+    return pattern.replaceAll("([.(){}+|^$\\[\\]\\\\])", "\\\\$1");
+  }
+
+  private static class PreviousMethodDetection {
+
+    List<StringConstant> previousPatterns = new ArrayList<>();
+
+    private PreviousMethodDetection scan(MethodInvocationTree parent) {
+      MethodInvocationTree childMethod = childMethodInvocation(parent);
+      while (childMethod != null) {
+        if (AUTHORIZE_REQUESTS_MATCHER.matches(childMethod)) {
+          break;
+        }
+        visitMethodInvocation(childMethod);
+        childMethod = childMethodInvocation(childMethod);
+      }
+      return this;
+    }
+
+    private void visitMethodInvocation(MethodInvocationTree methodInvocation) {
+      for (MethodMatcher invocationMatcher : METHOD_MATCHERS) {
+        if (invocationMatcher.matches(methodInvocation)) {
+          previousPatterns.addAll(antMatchersPatterns(methodInvocation));
+        }
+      }
+    }
+
+    @CheckForNull
+    private static MethodInvocationTree childMethodInvocation(MethodInvocationTree parent) {
+      Tree methodSelect = parent.methodSelect();
+      if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+        ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
+        if (expression.is(Tree.Kind.METHOD_INVOCATION)) {
+          return (MethodInvocationTree) expression;
+        }
+      }
+      return null;
+    }
+
+  }
+
+  private static List<StringConstant> antMatchersPatterns(MethodInvocationTree mit) {
+    List<StringConstant> argumentsAsString = mit.arguments().stream()
+      .map(StringConstant::of)
+      .collect(Collectors.toList());
+    return argumentsAsString.stream().allMatch(Objects::nonNull) ? argumentsAsString : Collections.emptyList();
```

Comment:
Why only considering the cases having only literals?

I think we should also consider the following cases, as we already know they are wrong, even without considering `myVar`:
```java
  http.authorizeRequests().antMatchers("/admin/**", myVar, "/admin/login"); // Noncompliant
```

Therefore, I would simply skip the null values.

Reply:
It was a shortcut to exclude this method signature: .antMatchers(HttpMethod.GET,"/home")
And yes, in the case of antMatchers("/a", myVar, "/b"); we have some false-negative.
I'm going to check the method signature (methodSymbol.parameterTypes().get(0).is("java.lang.String")) instead, to not have false-positive in this case.